File: go/0020/solution.go
package solution

func isValid(s string) bool {
	dict := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}
	stack := make([]rune, 0)

	for _, char := range s {
		if char == '(' || char == '[' || char == '{' {
			stack = append(stack, char)
		} else {
			if len(stack) == 0 {
				return false
			}
			x := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if dict[char] != x {
				return false
			}
		}
	}

	return len(stack) == 0
}

File: cpp/0020/solution.cc
#include "solution.h"

#include <map>
#include <stack>

using namespace std;

bool Solution::isValid(string s) {
  stack<char> stack;
  map<char, char> dict = {
      {')', '('},
      {']', '['},
      {'}', '{'},
  };

  for (char el : s) {
    if (dict.count(el)) {
      if (stack.empty())
        return false;
      if (dict[el] != stack.top())
        return false;
      stack.pop();
    } else {
      stack.push(el);
    }
  }

  return stack.empty();
}
